source/extensions/filters/http/tap/tap_config_impl.cc

```diff
@@ -177,8 +177,10 @@ bool HttpPerRequestTapperImpl::onDestroyLog() {
 void HttpPerRequestTapperImpl::onBody(
     const Buffer::Instance& data, Extensions::Common::Tap::TraceWrapperPtr& buffered_streamed_body,
     uint32_t max_buffered_bytes, MutableBodyChunk mutable_body_chunk,
-    MutableMessage mutable_message) {
-  // TODO(mattklein123): Body matching.
+    MutableMessage mutable_message, bool request) {
+  // Invoke body matcher.
+  request ? config_->rootMatcher().onRequestBody(data, statuses_)
```

Comment:
so I'll admit unfamiliarity with the tap filter, but I think this might not work with streaming data?
it's common that we'd have body data hit the tap filter in several chunks.  Imagine an 
onData("dead", false)
onData("beef", false)
when looking for deadbeef.  
I think this will fail with the current code unless I'm missing something.

Stepping back, we can ask the HCM to buffer data until we've hit the buffer limit, or we could possibly check the buffered_full_trace_.  Either way we're still going to have an implicit limit (because infinite buffering is an OOM vector) and it'd be good to both document what that limit is in the API, what happens if we go over that limit, and add an integration test where we intentionally send the string we're looking for in multiple chunks via codec_client_->sendData() ... upstream_request_->waitForData()) to make sure it gets picked up

Reply:
@alyssawilk Thanks for catching this. I will try to code it such way that multiple calls to onBody are treated as one logical unit vs separate buffers.

Reply:
To follow up on this. I modified matcher's code that it scans incoming body data chunks without buffering it. The code only buffers as many bytes as is the longest pattern which needs to be located - 1. So for example for string "envoyproxy" which is 10 chars long, the code will buffer 9 bytes to catch extreme cases when the string is spread across 2 body chunks "envoyprox"/"y". To be exact the code works when the pattern is spread across multiple body chunks: "envo"/"yprox"/"y".